src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,601 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Network Data Publisher.
+ *
+ */
+
+#include "network_data_publisher.hpp"
+
+#if OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "common/random.hpp"
+#include "thread/network_data_local.hpp"
+#include "thread/network_data_service.hpp"
+
+namespace ot {
+namespace NetworkData {
+
+//---------------------------------------------------------------------------------------------------------------------
+// Publisher
+
+Publisher::Publisher(Instance &aInstance)
+    : InstanceLocator(aInstance)
+#if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+    , mDnsSrpServiceEntry(aInstance)
+#endif
+    , mTimer(aInstance, Publisher::HandleTimer)
+{
+}
+
+void Publisher::HandleNotifierEvents(Events aEvents)
+{
+    OT_UNUSED_VARIABLE(aEvents);
+
+    bool registerWithLeader = false;
+
+#if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+    if (mDnsSrpServiceEntry.HandleNotifierEvents(aEvents))
+    {
+        registerWithLeader = true;
+    }
+#endif
+
+    if (registerWithLeader)
```

Comment:
Are we planning future enhancements that do not require `OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE`? If so, maybe add a comment here?

Reply:
Yes. For on-mesh prefix and external route entries. 

I have the implementation to add support for those and was originally planning to add it as a separate PR but I think would be better to update this PR itself (include everything in one commit). Will add this in next push.